integrations/langfuse/src/haystack_integrations/tracing/langfuse/tracer.py

```diff
@@ -137,6 +143,15 @@ def trace(self, operation_name: str, tags: Optional[Dict[str, Any]] = None) -> I
 
         span.raw_span().end()
         self._context.pop()
+
+        if len(self._context) == 1:
+            # The root span has to be a trace, which need to be removed from the context after the pipeline run
+            self._context.pop()
```

Comment:
Can you explain why this is needed?

Reply:
At the end of the pipeline.run, we still have the root trace in the context. All other pipeline runs will create spans under this root span. So when I run now a web server and executing the run method in every request I will get I will end up with one single trace and all requests will be under this trace.

e.g. The current implementation will store every pipeline run under the same trace:

trace: "Haystack" runtime 24 minutes (trace_id: 1234)
   span: pipeline.run: 10 seconds
     span: retrieval 3 seconds
     span: llm 7 seconds
   span: pipeline.run: 11 seconds
     span: retrieval 4 seconds
     span: llm 7 seconds
   ....

instead of bundling each run in its own trace

trace: "Haystack" runtime 10 seconds (trace_id: 1234)
   span: pipeline.run: 10 seconds
     span: retrieval 3 seconds
     span: llm 7 seconds

trace: "Haystack" runtime 11 seconds (trace_id: 5678)
   span: pipeline.run: 11 seconds
     span: retrieval 4 seconds
     span: llm 7 seconds

....

Because the Langfuse tracer is returning the **Trace_id** it is much better to a unique trace id for each request than returning always the same trace ID for every request within the same session.

We would be then able to add scores to each pipeline run using the langfuse sdk https://langfuse.com/docs/scores/custom.

Reply:
Gotcha, thanks for the explanation